src/expressions/match-expr.md

```diff
@@ -125,6 +126,15 @@ let message = match maybe_digit {
 > assert_eq!(i.get(), 2);
 > ```
 
+A pattern guard may refer to the variables bound within the pattern they follow.
+When such a variable's binding mode is by-value,
```

Comment:
Hmm... using "by-reference" here sounds like the variable is of a reference type when used inside the arm's body, which feels confusing. How about?

> A pattern guard may refer to the variables bound within the pattern they follow. For a by-value variable, before evaluating the guard, a shared reference is taken to the part of the scrutinee the variable matches on. While evaluating the guard, this shared reference is then used when accessing the variable. Only when the guard evaluates to true is the value moved, or copied, into the by-value variable. This allows shared borrows to be used inside guards without moving out of the scrutinee in case guard fails to match.

Reply:
That looks good. What I meant by the suggestion is that the following statement is true:

> When such a variable's binding mode is by-reference, a shared reference is taken to it before evaluating the guard.

This is observable only for mutable references, where it ensure that one cannot mutate through that reference. It might make more sense to put this in the part about mutation in guards.

Reply:
Ah, makes sense. I dropped the "by-value" part so that the paragraph applies to by-reference bindings as well and elaborated on preventing mutation.

Reply:
There's more that could be said about mutation in guard, but it's not really related to `bind_by_move_pattern_guards`. This is fine for now.